### PR TITLE
Implement file path sanitation

### PR DESCRIPTION
**Pull request by henrybear327**
As mentioned in issue #137, the elf path is not sanitized before opening, which might be a security risk. This PR addresses this issue by porting the path sanitation logic from Golang to C.

Reference code: https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/path/path.go;l=70

Unit test -> https://github.com/sysprog21/rv32emu/pull/274

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -67,3 +69,99 @@
     tp->tv_sec = tv_sec;
     tp->tv_nsec = tv_usec / 1000; /* Transfer to microseconds */
 }
+
+char *sanitize_path(const char *orig_path)
+{
+    size_t n = strlen(orig_path);
```
[the remark this answers is not in the document]
Comment (henrybear327) [reply src/utils.c]: I have no idea how we can approach this, as this is a command-line argument.

File: src/utils.c [hunk 2]
```diff
@@ -67,3 +69,99 @@
...
+
+    char *ret = (char *) malloc(n + 1);
+    memset(ret, '\0', n + 1);
```
[the remark this answers is not in the document]
Comment (henrybear327) [reply src/utils.c]: It's not a secure data.

File: src/utils.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jserv) [reply src/utils.c]: How about using [strnlen](https://man7.org/linux/man-pages/man3/strnlen.3.html)?

File: src/utils.c [hunk 3]
```diff
@@ -67,3 +69,98 @@ void rv_clock_gettime(struct timespec *tp)
     tp->tv_sec = tv_sec;
     tp->tv_nsec = tv_usec / 1000; /* Transfer to microseconds */
 }
+
+char *sanitize_path(const char *orig_path)
+{
+    size_t n = strlen(orig_path);
+
+    char *ret = (char *) malloc(n + 1);
```
Comment (jserv) [inline src/utils.c]: In C, there is no need to cast. i.e., remove `(char *) `.

**Review by jserv — CHANGES_REQUESTED** [review]
Merge the proposed changes in #274

File: src/utils.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (henrybear327) [reply src/utils.c]: Cap the length at 1024 for now.

**Comment by henrybear327** [comment]
> Merge the proposed changes in #274

Merged using a separate commit.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch where I revert Codacy to avoid unnecessary MISRA C checks.

File: src/elf.c [hunk 4]
```diff
@@ -290,16 +291,20 @@ bool elf_load(elf_t *e, riscv_t *rv, memory_t *mem)
     return true;
 }
 
-bool elf_open(elf_t *e, const char *path)
+bool elf_open(elf_t *e, const char *_path)
```
Comment (jserv) [inline src/elf.c]: Don't use variable name starting with underscore. Read `CONTRIBUTING.md` carefully.

File: src/main.c [hunk 5]
```diff
@@ -188,6 +189,10 @@ static void dump_test_signature(elf_t *elf)
 
 int main(int argc, char **args)
 {
+#ifdef UNITTEST
```
Comment (jserv) [inline src/main.c]: Please avoid including the `UNITTEST` macro in any files located within the `src/` directory.

Refrain from including unit test code within the `src/` directory. Instead, utilize mock functions. These functions serve to facilitate testing by replacing the actual implementation of a function, recording calls to that function (along with the parameters used), capturing constructor function instances, and permitting configuration of return values for testing purposes. The unit test programs based on the mock functions should be placed in the `tests/` directory.

File: src/utils.c [hunk 6]
```diff
@@ -67,3 +72,159 @@ void rv_clock_gettime(struct timespec *tp)
...
+                w++;
+            }
+
+            /* copy element */
+            for (; r < n && orig_path[r] != '/'; r++) {
+                ret[w] = orig_path[r];
+                w++;
+            }
+        }
+        // printf("w = %ld, r = %ld, dotdot = %ld\nret = %s\n", w, r, dotdot,
```
Comment (jserv) [inline src/utils.c]: Don't leave dead code!

File: src/utils.c [hunk 7]
```diff
@@ -67,3 +72,159 @@ void rv_clock_gettime(struct timespec *tp)
...
+        ret[i] = '\0';
+    }
+    return ret;
+}
+
+#ifdef UNITTEST
+void compare(char *input, char *expected_output)
+{
+    char *input_sanitized = sanitize_path(input);
+    // printf("\n\nInput =\t\t\t%s\nOutput =\t\t%s\nExpected output =\t%s\n",
```
Comment (jserv) [inline src/utils.c]: Ditto. Always make the code clean.

**Review by jserv — CHANGES_REQUESTED** [review]
Use C-style comments. i.e., `/* comments */`.

**Review by jserv — CHANGES_REQUESTED** [review]
Check `mk/tests.mk` for existing unit test programs along with mock functions.

File: src/utils.c (hunk 7 above)
Comment (henrybear327) [reply src/utils.c]: Improved

File: src/elf.c [hunk 8]
```diff
@@ -290,16 +291,20 @@ bool elf_load(elf_t *e, riscv_t *rv, memory_t *mem)
     return true;
 }
 
-bool elf_open(elf_t *e, const char *path)
+bool elf_open(elf_t *e, const char *orig_path)
```
Comment (jserv) [inline src/elf.c]: Rename from `orig_path` to `input` or simpler `in`.

File: tests/path/test-path.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "utils.h"
+
+void compare(char *input, char *expected_output)
+{
+    char *input_sanitized = sanitize_path(input);
+    if (strcmp(input_sanitized, expected_output) != 0) {
```
Comment (jserv) [inline tests/path/test-path.c]: Remove `!= 0`.

File: mk/tools.mk [hunk 10]
```diff
@@ -7,6 +7,7 @@ HIST_OBJS := \
 	elf.o \
 	decode.o \
 	mpool.o \
+	utils.o \
```
Comment (jserv) [inline mk/tools.mk:10]: Don't add `utils.o`.

File: src/utils.c [hunk 11]
```diff
@@ -67,3 +71,98 @@ void rv_clock_gettime(struct timespec *tp)
...
+    }
+
+    int rooted = (input[0] == '/');
+
+    /*
+     * Invariants:
+     *	reading from path; r is index of next byte to process -> path[r]
+     *	writing to buf; w is index of next byte to write -> ret[strlen(ret)]
+     *	dotdot is index in buf where .. must stop, either because
+     *		a) it is the leading slash
```
Comment (jserv) [inline src/utils.c]: Indent.

File: src/utils.c [hunk 12]
```diff
@@ -67,3 +71,98 @@ void rv_clock_gettime(struct timespec *tp)
...
+    }
+
+    /* Turn empty string into "." */
+    if (w == 0) {
+        ret[w] = '.';
+        w++;
+    }
+
+    for (size_t i = w; i < n; i++) {
+        ret[i] = '\0';
```
Comment (jserv) [inline src/utils.c]: Why not using `memset`, which is usually optimized by libc?

File: tests/path/test-path.c [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+        exit(1);
+    }
+
+    free(input_sanitized);
+}
+
+void sanitize_path_test()
+{
+    // Already clean
```
Comment (jserv) [inline tests/path/test-path.c]: C-style comments!

File: src/utils.c (hunk 12 above)
Comment (henrybear327) [reply src/utils.c]: Not properly thought of :( Implemented with `memset` now

File: tests/path/test-path.c (hunk 13 above)
Comment (henrybear327) [reply tests/path/test-path.c]: I think I will need some more days to get my brain used to not use `//` :'(

File: src/utils.c [hunk 14]
```diff
@@ -21,6 +23,8 @@
 #endif
 #endif
 
+const int MAX_PATH_LEN = 1024;
```
Comment (jserv) [inline src/utils.c]: Use macro or `enum` instead. In C, an object with const integer type occupies extra memory space.

Comment (henrybear327) [reply src/utils.c]: Learned and changed.

File: tests/path/test-path.c [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    compare("abc//./../def", "def");
+    compare("abc/../../././../def", "../../def");
+}
+
+int main()
+{
+    sanitize_path_test();
+
+    return 0;
+}
```
Comment (jserv) [inline tests/path/test-path.c]: See https://medium.com/@alexey.inkin/how-to-force-newline-at-end-of-files-and-why-you-should-do-it-fdf76d1d090e

**Review by jserv — CHANGES_REQUESTED** [review]
Enforce newline at end of files.

File: tests/path/test-path.c (hunk 15 above)
Comment (henrybear327) [reply tests/path/test-path.c]: I am aware of this but sorry for missing out on this issue :(

File: src/utils.c [hunk 16]
```diff
@@ -67,3 +71,96 @@ void rv_clock_gettime(struct timespec *tp)
...
+{
+    size_t n = strnlen(input, MAX_PATH_LEN);
+
+    char *ret = malloc(MAX_PATH_LEN + 1);
+    memset(ret, '\0', MAX_PATH_LEN + 1);
+
+    /* After sanitization, the new path will only be shorter than the original
+     * one. Thus, we can reuse the space */
+    if (n == 0) {
+        ret[0] = '.';
```
Comment (jserv) [inline src/utils.c:88]: If `n` equals 0, there is no need to allocate space larger than `MAX_PATH_LEN` for `ret`, correct? I expect a memory-efficient approach.

File: src/utils.c [hunk 17]
```diff
@@ -67,3 +71,96 @@ void rv_clock_gettime(struct timespec *tp)
...
+        } else if (input[r] == '.' && (r + 1 == n || input[r + 1] == '/')) {
+            /* . element */
+            r++;
+        } else if (input[r] == '.' && input[r + 1] == '.' &&
+                   (r + 2 == n || input[r + 2] == '/')) {
+            /* .. element: remove to last / */
+            r += 2;
+
+            if (w > dotdot) {
+                /* can backtrack */
```
Comment (jserv) [inline src/utils.c:125]: should be "backtrace"?

File: src/utils.c [hunk 18]
```diff
@@ -67,3 +71,96 @@ void rv_clock_gettime(struct timespec *tp)
...
+                    w++;
+                }
+                ret[w] = '.';
+                w++;
+                ret[w] = '.';
+                w++;
+                dotdot = w;
+            }
+        } else {
+            /* real path element.
```
Comment (jserv) [inline src/utils.c]: Use single line comment.

File: src/utils.c (hunk 16 above)
Comment (henrybear327) [reply src/utils.c:88]: Improved

File: src/utils.h [hunk 19]
```diff
@@ -21,3 +21,31 @@ void rv_clock_gettime(struct timespec *tp);
...
+ * If the result of this process is an empty string, Clean
+ * returns the string ".".
+ *
+ * See also Rob Pike, “Lexical File Names in Plan 9 or
+ * Getting Dot-Dot Right,”
+ * https://9p.io/sys/doc/lexnames.html
+ */
+char *sanitize_path(const char *input);
+
+void sanitize_path_test();
```
Comment (jserv) [inline src/utils.h]: Remove `sanitize_path_test`.

File: src/utils.c (hunk 17 above)
Comment (henrybear327) [reply src/utils.c:125]: It's taken from the Golang codebase. 

I believe the original meaning was indeed backtracking, as we are moving backward on the path string from the current position of `w` (e.g. `/a/b/...` will be backtracked to `a`). 

Backtrace is a stack trace, probably not what we want here IMO.

**Comment by jserv** [comment]
Is it time to close #137? If so, append `Close #137` at the end of git commit message.

**Comment by henrybear327** [comment]
> Is it time to close #137? If so, append `Close #137` at the end of git commit message.

Yes. Commit messages changed as requested

File: src/utils.c [hunk 20]
```diff
@@ -67,3 +71,95 @@ void rv_clock_gettime(struct timespec *tp)
...
+    memset(ret, '\0', n + 1);
+
+    /* After sanitization, the new path will only be shorter than the original
+     * one. Thus, we can reuse the space */
+    if (n == 0) {
+        ret[0] = '.';
+        return ret;
+    }
+
+    int rooted = (input[0] == '/');
```
Comment (jserv) [inline src/utils.c]: Use `bool` for `rooted`. Or, rename it to `is_root`.

File: src/utils.c [hunk 21]
```diff
@@ -67,3 +71,95 @@ void rv_clock_gettime(struct timespec *tp)
     tp->tv_sec = tv_sec;
     tp->tv_nsec = tv_usec / 1000; /* Transfer to microseconds */
 }
+
+char *sanitize_path(const char *input)
+{
+    size_t n = strnlen(input, MAX_PATH_LEN);
+
+    char *ret = malloc(n + 1);
+    memset(ret, '\0', n + 1);
```
Comment (jserv) [inline src/utils.c]: Use `calloc` instead of `malloc` + `memset`.
See https://vorpus.org/blog/why-does-calloc-exist/

Comment (henrybear327) [reply src/utils.c]: Thanks for the reference

**Pull request merged**